File: 백준/Gold/14502. 연구소/연구소.py
from collections import deque

dxs = [1, 0, -1, 0]
dys = [0, 1, 0, -1]

def is_reachable(x, y):
    return 0 <= x < n and 0 <= y < m

def bfs(arr):
    temp_cnt = 0

    pq = deque(arr)

    for x, y in arr:
        visited[x][y] = True

    while pq:
        x, y = pq.popleft()

        for dx, dy in zip(dxs, dys):
            nx, ny = x + dx, y + dy

            if is_reachable(nx, ny) and matrix[nx][ny] == 0 and not visited[nx][ny]:
                visited[nx][ny] = True
                matrix[nx][ny] = 2
                pq.append([nx, ny])

    safe_count = sum(row.count(0) for row in matrix)
    return safe_count
                

n, m = map(int, input().split())

matrix_ = []
for _ in range(n):
    matrix_.append(list(map(int, input().split())))

empty_loc = []

for i in range(n):
    for j in range(m):
        if matrix_[i][j] == 0:
            empty_loc.append([i, j])

max_cnt = 0

for i in range(len(empty_loc)):
    for j in range(i + 1, len(empty_loc)):
        for k in range(j + 1, len(empty_loc)):
            matrix = [x[:] for x in matrix_]

            x1, y1 = empty_loc[i]
            x2, y2 = empty_loc[j]
            x3, y3 = empty_loc[k]

            matrix[x1][y1] = 1; matrix[x2][y2] = 1; matrix[x3][y3] = 1

            visited = [[False] * m for _ in range(n)]

            temp_arr = []

            for x in range(n):
                for y in range(m):
                    if matrix[x][y] == 2:
                        temp_arr.append([x, y])

            max_cnt = max(max_cnt, bfs(temp_arr))


print(max_cnt)

File: 백준/Gold/14502. 연구소/연구소.cc
#include <bits/stdc++.h>
using namespace std;

int arr[9][9];
int n, m, temp, ans;
vector<pair<int, int>> walls;
vector<pair<int, int>> viruses;
bool visited[9][9];

const int dxs[4] = {1, 0, -1, 0};
const int dys[4] = {0, 1, 0, -1};

bool is_reachable(int x, int y) {
	return (0 <= x && x < n) && (0 <= y && y < m);
}

void bfs(int x, int y) {
	queue<pair<int, int>> q;

	visited[x][y] = true;
	q.push({x, y});

	while (q.size()) {
		tie(x, y) = q.front(); q.pop();

		for (int i = 0; i < 4; ++i) {
			int nx = x + dxs[i];
			int ny = y + dys[i];

			if (is_reachable(nx, ny) && arr[nx][ny] == 0 && !visited[nx][ny]) {
				visited[nx][ny] = true;
				q.push({nx, ny});
			}
		}
	}
}

int bfs_all() {
	int cnt = 0;

	memset(visited, false, sizeof(visited));

	for (pair<int, int> virus : viruses) {
		bfs(virus.first, virus.second);
	}

	for (int i = 0; i < n; ++i) {
		for (int j = 0; j < m; ++j) {
			if (arr[i][j] == 0 && !visited[i][j]) cnt++;
		}
	}

	return cnt;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
	
	cin >> n >> m;

	for (int i = 0; i < n; ++i) {
		for (int j = 0; j < m; ++j) {
			cin >> temp;
			if (temp == 0) walls.push_back({i, j});
			if (temp == 2) viruses.push_back({i, j});
			arr[i][j] = temp;
		}
	}

	for (int i = 0; i < walls.size(); ++i) {
		for (int j = i + 1; j < walls.size(); ++j) {
			for (int k = j + 1; k < walls.size(); ++k) {
				arr[walls[i].first][walls[i].second] = 1;
				arr[walls[j].first][walls[j].second] = 1;
				arr[walls[k].first][walls[k].second] = 1;

				ans = max(ans, bfs_all());

				arr[walls[i].first][walls[i].second] = 0;
				arr[walls[j].first][walls[j].second] = 0;
				arr[walls[k].first][walls[k].second] = 0;
			}
		}
	}

	cout << ans;

	return 0;
}